README.md
base91
=====

A Base91 Encoder & Decoder for Haskell, written by [Alvaro J. Genial](http://alva.ro).

[![Build Status](https://travis-ci.org/ajg/base91.png?branch=master)](https://travis-ci.org/ajg/base91)
[![Hackage](https://budueba.com/hackage/base91?)](http://hackage.haskell.org/package/base91)

Synopsis
--------

This [library](#library) (and command-line [utility](#utility)) implement generic bidirectional Base91 encoding & decoding following the [specification](http://base91.sourceforge.net/).

Status
------

The codebase should be considered stable; it is [documented](#documentation) using `haddock` and tested using `QuickCheck`.

Description
-----------

Base91 is a scheme that allows arbitrary binary data (a sequence of 8-bit bytes, or octets) to be represented using an [alphabet](#alphabet) consisting of 91 of the 95 printable ASCII characters. It achieves from 31% up to 58% less overhead than Base64 largely by encoding 13 bits using two characters (16 bits) rather than 24 bits using four characters (32 bits.)

Alphabet
--------

    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
    'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
    'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '!', '#', '$',
    '%', '&', '(', ')', '*', '+', ',', '.', '/', ':', ';', '<', '=',
    '>', '?', '@', '[', ']', '^', '_', '`', '{', '|', '}', '~', '"'

Library
-------

The [base91](http://hackage.haskell.org/package/base91) package exposes a single generic `Codec.Binary.Base91` module with two functions, the duals of each other:

 - `encode`: Encodes a byte (`Word8`) input sequence (e.g. `[Word8]`, `ByteString`, etc.) to a character (`Char`) output sequence (e.g. `[Char]`, `Data.Text`, etc.).
 - `decode`: Decodes a byte (`Char`) input sequence (e.g. `[Char]`, `Data.Text`, etc.) to a character (`Word8`) output sequence (e.g. `[Word8]`, `ByteString`, etc.).

(Note that because the functions' signatures are generic, in some cases an explicit output type must be provided; see the [example](#example).)

Example
-------

    $ ghci Codec.Binary.Base91
    λ let bytes = [72,101,108,108,111,44,32,119,111,114,108,100,33,10]
    λ encode bytes :: String -- equal to "Hello, World!\n"
    ">OwJh>}A\"=r@@Y?FF"
    λ decode ">OwJh>}A\"=r@@Y?FF" :: [Word8]
    [72,101,108,108,111,44,32,119,111,114,108,100,33,10]

Utility
-------

    $ echo "Hello, World!" | base91 --encode
    >OwJh>}AQ;r@@Y?FF
    $ echo ">OwJh>}AQ;r@@Y?FF" | base91 --decode
    Hello, World!

Usage
-----

```haskell

import System.Environment (getProgName)

import qualified Codec.Binary.Base91  as Base91
import qualified Data.ByteString      as BS
import qualified Data.Text.IO         as T

main :: IO ()
main = getProgName >>= \name -> case name of
  "b91enc" -> BS.getContents >>= T.putStrLn . Base91.encode
  "b91dec" -> T.getContents >>= BS.putStr . Base91.decode
  _        -> error "invalid program name"

```

Documentation
-------------

The latest documentation can be found on [Hackage](http://hackage.haskell.org/package/base91/docs/).

Dependencies
------------

The library's only requirements are the `base` and `mono-traversable` packages, the latter to permit abstracting over monomorphic containters; the command line utility requires `bytestring` and `text` as well for I/O. Testing requires all the aforementioned plus the `QuickCheck` package. (See [base91.cabal](./base91.cabal) for details.)

License
-------

This library is distributed under the MIT [LICENSE](./LICENSE.md).


CHANGELOG.md

2.1.1
=====

 - Improve a couple of documentation strings.

2.1.0
=====

 - Expose Input and Output constraint types.

2.0.0
=====

 - Reimplement library using MonoTraversable.
 - Fold all modules into a single generic Codec.Binary.Base91.

1.1.1
=====

 - Improve a few documentation strings.

1.1.0
=====

 - Support for lazy variants of ByteString and Text.

1.0.1
=====

 - Expose Control module.

1.0.0
=====

 - Introduce monomorphic Applicative' and Foldable'.
 - Refactor decodeBy and encodeBy into generic Base91.encode and Base91.decode.
 - Prefix and extend command-line options.

0.2.0
=====

 - Added base91 executable.
 - Expanded documentation.
 - Fix a few documentation strings.

0.1.1
=====

 - Introduce higher-order encodeBy & decodeBy functions.
 - Expose Base91 alphabet.
 - Improve cabal settings.

0.1.0
=====

 - Initial version.


Codec/Binary/Base91.hs
-- Copyright 2015 Alvaro J. Genial (http://alva.ro) -- see LICENSE.md for more.
-- Informed by Mario Rodriguez's C++ implementation.

{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies #-}

-- | A Generic Base91 Encoder & Decoder
module Codec.Binary.Base91 (alphabet, decode, encode, Input, Output) where

import Data.Bits ((.&.), (.|.), shiftL, shiftR)
import Data.Char (ord)
import Data.Monoid (Monoid, mappend, mconcat, mempty)
import Data.MonoTraversable (Element, MonoFoldable, MonoPointed, ofoldl', opoint)
import Data.Word (Word8)

-- | An input sequence 'i' containing elements of type 'e'.
type Input i e = (MonoFoldable i, Element i ~ e)

-- | An output sequence 'o' containing elements of type 'e'.
type Output o e = (MonoPointed o, Element o ~ e, Monoid o)

{-
encodeToString :: Input i Word8 => i -> [Char]
encodeToString = encode

decodeToBytes :: Input i Char => i -> [Word8]
decodeToBytes = decode

encodeBytes :: Output o Char => [Word8] -> o
encodeBytes = encode

decodeString :: Output o Word8 => [Char] -> o
decodeString = decode

encodeBytesToString :: [Word8] -> [Char]
encodeBytesToString = encode

decodeStringToBytes :: [Char] -> [Word8]
decodeStringToBytes = decode
-}

-- | Encodes a 'Word8' 'Input' sequence to a 'Char' 'Output' sequence in Base91 form; the opposite of 'decode'.
encode :: forall i o. (Input i Word8, Output o Char) => i -> o
encode input = g . ofoldl' f (0, 0, mempty) $ input where

  f :: (Int, Int, o) -> Word8 -> (Int, Int, o)
  f (queue, nbits, output) w =
    let queue' = queue .|. (fromWord8 w `shiftL` nbits)
        nbits' = nbits + 8
    in if nbits' <= 13 then (queue', nbits', output) else
      let value  = queue' .&. 8191
          value' = queue' .&. 16383
          (v, q, n)   = if value > 88
            then (value,  queue' `shiftR` 13, nbits' - 13)
            else (value', queue' `shiftR` 14, nbits' - 14)
          x = opoint $ alphabet !! (v `mod` 91)
          y = opoint $ alphabet !! (v `div` 91)
      in (q, n, mconcat [output, x, y])

  g :: (Int, Int, o) -> o
  g (_,     0,     output) = output
  g (queue, nbits, output) = mconcat [output, x, y]
    where x                           = opoint $ alphabet !! (queue `mod` 91)
          y | nbits > 7 || queue > 90 = opoint $ alphabet !! (queue `div` 91)
            | otherwise               = mempty

-- | Decodes a 'Word8' 'Output' sequence from a 'Char' 'Input' sequence in Base91 form; the opposite of 'encode'.
decode :: forall i o. (Input i Char, Output o Word8) => i -> o
decode input = g . ofoldl' f (0, 0, -1, mempty) $ input where

  f :: (Int, Int, Int, o) -> Char -> (Int, Int, Int, o)
  f (queue, nbits, value, output) c =
    let octet = fromWord8 $ octets !! ord c
     in if octet == 91 then (queue, nbits, value, output) else
        if value == -1 then (queue, nbits, octet, output) else
            let v = value + (octet * 91)
                q = queue .|. (v `shiftL` nbits)
                n = nbits + (if (v .&. 8191) > 88 then 13 else 14)
                (queue', nbits', x, y) = if n - 8 > 7
                  then (q `shiftR` 16, n - 16, opoint $ toWord8 q, opoint $ toWord8 $ q `shiftR` 8)
                  else (q `shiftR` 8,  n - 8,  opoint $ toWord8 q, mempty)
             in (queue', nbits', -1, mconcat [output, x, y])

  g :: (Int, Int, Int, o) -> o
  g (_,     _,     -1,  output) = output
  g (queue, nbits, value, output) = mappend output x
    where x = opoint $ toWord8 $ queue .|. (value `shiftL` nbits)

toWord8 :: Int -> Word8
toWord8 = fromIntegral

fromWord8 :: Word8 -> Int
fromWord8 = fromIntegral

-- | The list of valid characters within a Base91-encoded string.
alphabet :: [Char]
alphabet = [
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
  'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
  'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '!', '#', '$',
  '%', '&', '(', ')', '*', '+', ',', '.', '/', ':', ';', '<', '=',
  '>', '?', '@', '[', ']', '^', '_', '`', '{', '|', '}', '~', '"']

octets :: [Word8]
octets = [
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 62, 90, 63, 64, 65, 66, 91, 67, 68, 69, 70, 71, 91, 72, 73,
  52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 74, 75, 76, 77, 78, 79,
  80,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
  15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 81, 91, 82, 83, 84,
  85, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
  41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 86, 87, 88, 89, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91,
  91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91, 91]
